examples/step-49/example.geo

```diff
@@ -1,56 +1,139 @@
+
+// Lines on the bottom/top domain
+Line(38) = {1, 37};
+Line(39) = {37, 2};
+Line(40) = {3, 36};
+Line(41) = {36, 4};
+
+// Lines that bisect the domain from left/right
+Line(42) = {37, 7};
+Line(43) = {7, 36};
+
+// Lines on the right hand side of the domain
+Line(52) = {2, 107};
+Line(53) = {107, 109};
+Line(54) = {109, 110};
+Line(55) = {110, 108};
+Line(56) = {108, 14};
+Line(57) = {14, 111};
+Line(58) = {112, 3};
+Line(59) = {111, 113};
+Line(60) = {113, 112};
+
+//---------------------------------------------------------------------------
+// Here are the points and lines that create the two holes in the domain
+
+// Relevant points for bottom left hole
+Point(31) = {-2.1, -1.54, 0, 1.0}; // origin
+Point(32) = {-2.1, -0.84, 0, 1.0}; // top point
+Point(33) = {-2.1, -2.24, 0, 1.0}; // bottom point 
+Point(103) = {-1.4, -1.54, 0, 1.0}; // right point
+Point(104) = {-2.8, -1.54, 0, 1.0}; // left point
+
+// Bottom left hole
+Circle(48) = {103, 31, 32};
+Circle(49) = {32, 31, 104};
+Circle(50) = {104, 31, 33};
+Circle(51) = {33, 31, 103};
+
+// The upper right circle
+radius1 = .5;
+Point(19) = {0.42, 2.0, 0, 1.0}; // origin 
+Point(22) = {0.42, 2.0 + radius1, 0, 1.0}; // top point
+Point(23) = {0.42, 2.0 - radius1, 0, 1.0}; // bottom point
+Circle(12) = {0.42, 2.0, 0, radius1, 0, 2*Pi};
+
+//---------------------------------------------------------------------------
+// This section describes the "Plane Surfaces", i.e., the 2D surfaces for meshing
+
+// The surface of the top right partition
+Curve Loop(1) = {40, -43, 14, 57, 59, 60, 58};
+Curve Loop(2) = {12};
+Plane Surface(1) = {1, 2};
+
+// Surface of the bottom right partition
+Curve Loop(3) = {39, 52, 53, 54, 55, 56, -14, -42};
+Plane Surface(2) = {3};
+
+// Surface of bottem left mesh
+Curve Loop(4) = {42, 7, 8, 9, 38};
+Curve Loop(5) = {49, 50, 51, 48};
+Plane Surface(3) = {4, 5};
+
+// Surface of top left mesh
+Curve Loop(6) = {6, 43, 41, 4, 5};
+Plane Surface(4) = {6};
+
+// Creates a physical surface.  The expression list on the right hand side is the list
+// of elementary surfaces created above.  This is what makes our mesh 2D.
+Physical Surface(2) = {1, 2, 3, 4};
+
+//---------------------------------------------------------------------------
+// This section describes the physical IDs of certain objects.
+
+// Assign boundary ID of 0 to outer boundary
+Physical Curve(0) = {38, 39, 52, 53, 54, 55, 56, 57, 59, 60, 58, 40, 41, 4, 5, 6, 7, 8, 9};
+
+// Assign boundary ID of 1 top right circle
+Physical Curve(1) = {12};
+
+// Assign boundary IDs of 2 & 3 to the top & bottom half of the bottom left
+Physical Curve(2) = {48, 49};
+Physical Curve(3) = {50, 51};
+
+//---------------------------------------------------------------------------
+// Parameters for the meshing
+
+Mesh.Algorithm = 3;
 Mesh.RecombineAll = 1;
-Mesh.CharacteristicLengthFactor = 0.09;
+Mesh.CharacteristicLengthFactor = .6;
 Mesh.SubdivisionAlgorithm = 1;
 Mesh.Smoothing = 20;
 Show "*";
```

Comment:
Unrelated, but maybe for future work you would like to demonstrate also how the mesh size can be set dynamically by defining refinement regions or attractors for the mesh size within GMSH. This is a nice feature that is not trivial at all to use. We can talk about it if you want.

Reply:
We can try to add a length scale for one of the points/lines/regions. Would that make sense?

Reply:
@tjhei Sorry for the delay. Yeah that would absolutely make sense. Another option is to use Mesh size field (see t10.geo https://gmsh.info/doc/texinfo/gmsh.html#t10), but maybe this would turn this step too much into a GMSH tutorial?